stan/math/rev/fun/trace_quad_form.hpp

```diff
@@ -14,83 +14,51 @@
 
 namespace stan {
 namespace math {
-namespace internal {
-template <typename Ta, int Ra, int Ca, typename Tb, int Rb, int Cb>
-class trace_quad_form_vari_alloc : public chainable_alloc {
- public:
-  trace_quad_form_vari_alloc(const Eigen::Matrix<Ta, Ra, Ca>& A,
-                             const Eigen::Matrix<Tb, Rb, Cb>& B)
-      : A_(A), B_(B) {}
 
-  double compute() { return trace_quad_form(value_of(A_), value_of(B_)); }
+template <typename EigMat1, typename EigMat2,
```

Comment:
Even though this has two arguments and could be broken out into three implementations (code [here](https://github.com/stan-dev/math/blob/4f3705c1386017faec72ebf852f0d1e754c3d863/stan/math/rev/fun/trace_quad_form.hpp)), I went with one.

Here are benchmarks broken into three different var/double input implementations (I don't know why the first is slow -- probably something that could be fixed):
```
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
toss_me                                   10067059 ns     10057209 ns           70
trace_quad_form/2/manual_time                  419 ns          724 ns      1663771
trace_quad_form/4/manual_time                  673 ns         1331 ns      1019910
trace_quad_form/8/manual_time                 1434 ns         3460 ns       491865
trace_quad_form/16/manual_time                5120 ns        13416 ns       134934
trace_quad_form/32/manual_time               30203 ns        68312 ns        23236
trace_quad_form/64/manual_time              213271 ns       381398 ns         3592
trace_quad_form/128/manual_time            1647797 ns      2342545 ns          376
trace_quad_form/256/manual_time           12940950 ns     16308948 ns           57
trace_quad_form/512/manual_time           91161131 ns    105007828 ns            6
trace_quad_form/1024/manual_time         815430043 ns    886466576 ns            1
trace_quad_form_data_A/2/manual_time           337 ns          620 ns      2154932
trace_quad_form_data_A/4/manual_time           688 ns         1455 ns      1156487
trace_quad_form_data_A/8/manual_time          1133 ns         2989 ns       540222
trace_quad_form_data_A/16/manual_time         3549 ns         9960 ns       153987
trace_quad_form_data_A/32/manual_time        19209 ns        46700 ns        37916
trace_quad_form_data_A/64/manual_time       128111 ns       240987 ns         5633
trace_quad_form_data_A/128/manual_time      948054 ns      1413701 ns          739
trace_quad_form_data_A/256/manual_time     8214071 ns     10240882 ns           86
trace_quad_form_data_A/512/manual_time    70229497 ns     78740468 ns           11
trace_quad_form_data_A/1024/manual_time  462998300 ns    510277851 ns            2
trace_quad_form_data_B/2/manual_time           353 ns          651 ns      2144927
trace_quad_form_data_B/4/manual_time           556 ns         1193 ns      1241467
trace_quad_form_data_B/8/manual_time          1021 ns         2807 ns       685590
trace_quad_form_data_B/16/manual_time         3629 ns        10400 ns       197400
trace_quad_form_data_B/32/manual_time        20944 ns        50909 ns        29866
trace_quad_form_data_B/64/manual_time       122996 ns       231928 ns         5894
trace_quad_form_data_B/128/manual_time      836762 ns      1260478 ns          792
trace_quad_form_data_B/256/manual_time     7263060 ns      9077717 ns           92
trace_quad_form_data_B/512/manual_time    55530521 ns     63071147 ns           12
trace_quad_form_data_B/1024/manual_time  424628951 ns    467905904 ns            2
```

Here are benchmarks for the single implementation:
```
----------------------------------------------------------------------------------
Benchmark                                        Time             CPU   Iterations
----------------------------------------------------------------------------------
toss_me                                   10728582 ns     10738756 ns           50
trace_quad_form/2/manual_time                  498 ns          869 ns      1720492
trace_quad_form/4/manual_time                  739 ns         1481 ns       916035
trace_quad_form/8/manual_time                 1634 ns         3890 ns       425756
trace_quad_form/16/manual_time                5417 ns        14142 ns       133363
trace_quad_form/32/manual_time               28244 ns        64834 ns        24729
trace_quad_form/64/manual_time              177849 ns       320618 ns         3782
trace_quad_form/128/manual_time            1303900 ns      1881009 ns          527
trace_quad_form/256/manual_time           11417925 ns     14418493 ns           60
trace_quad_form/512/manual_time           85913338 ns     99149686 ns            7
trace_quad_form/1024/manual_time         691839532 ns    761973120 ns            1
trace_quad_form_data_A/2/manual_time           306 ns          565 ns      2316540
trace_quad_form_data_A/4/manual_time           504 ns         1051 ns      1349713
trace_quad_form_data_A/8/manual_time          1008 ns         2676 ns       695297
trace_quad_form_data_A/16/manual_time         3438 ns         9724 ns       203602
trace_quad_form_data_A/32/manual_time        18199 ns        44801 ns        38201
trace_quad_form_data_A/64/manual_time       118339 ns       224733 ns         5928
trace_quad_form_data_A/128/manual_time      855046 ns      1283016 ns          815
trace_quad_form_data_A/256/manual_time     8964877 ns     11102297 ns           91
trace_quad_form_data_A/512/manual_time    65373202 ns     73202014 ns           10
trace_quad_form_data_A/1024/manual_time  547440607 ns    591318001 ns            1
trace_quad_form_data_B/2/manual_time           329 ns          605 ns      2215157
trace_quad_form_data_B/4/manual_time           502 ns         1062 ns      1365079
trace_quad_form_data_B/8/manual_time           931 ns         2611 ns       751464
trace_quad_form_data_B/16/manual_time         3356 ns         9701 ns       207811
trace_quad_form_data_B/32/manual_time        17583 ns        43897 ns        39979
trace_quad_form_data_B/64/manual_time       117260 ns       223372 ns         5960
trace_quad_form_data_B/128/manual_time      944731 ns      1409704 ns          739
trace_quad_form_data_B/256/manual_time     7431989 ns      9315718 ns           87
trace_quad_form_data_B/512/manual_time    56982713 ns     64847214 ns           12
trace_quad_form_data_B/1024/manual_time  444922367 ns    490421169 ns            2
```